### PR TITLE
Using reddis cache for reads

### PR DESCRIPTION
**Pull request by Moro-Code**
# Summary | Résumé

This PR adds the usage of redis cache to retrieve template and service objects rather than re-query the database in `save_sms` and `save_email` thus reducing load on readers. 

# Help requested | Aide requise

Please take a careful look at the code. Are there any considerations that maybe I have not thought of. Are the tests covering everything. Is there more tests that need to be added ?

# Reviewer checklist | Liste de vérification du réviseur

This is a suggested checklist of questions reviewers might ask during their
review | Voici une suggestion de liste de vérification comprenant des questions
que les réviseurs pourraient poser pendant leur examen :

- [ ] Does this meet a user need? | Est-ce que ça répond à un besoin utilisateur?
- [ ] Is it accessible? | Est-ce que c’est accessible?
- [ ] Is it translated between both offical languages? | Est-ce dans les deux
      langues officielles?
- [ ] Is the code maintainable? | Est-ce que le code peut être maintenu?
- [ ] Have you tested it? | L’avez-vous testé?
- [ ] Are there automated tests? | Y a-t-il des tests automatisés?
- [ ] Does this cause automated test coverage to drop? | Est-ce que ça entraîne
      une baisse de la quantité de code couvert par les tests automatisés?
- [ ] Does this break existing functionality? | Est-ce que ça brise une
      fonctionnalité existante?
- [ ] Should this be split into smaller PRs to decrease change risk? | Est-ce
      que ça devrait être divisé en de plus petites demandes de tirage (« pull
      requests ») afin de réduire le risque lié aux modifications?
- [ ] Does this change the privacy policy? | Est-ce que ça entraîne une
      modification de la politique de confidentialité?
- [ ] Does this introduce any security concerns? | Est-ce que ça introduit des
      préoccupations liées à la sécurité?
- [ ] Does this significantly alter performance? | Est-ce que ça modifie de
      façon importante la performance?
- [ ] What is the risk level of using added dependencies? | Quel est le degré de
      risque d’utiliser des dépendances ajoutées?
- [ ] Should any documentation be updated as a result of this? (i.e. README
      setup, etc.) | Faudra-t-il mettre à jour la documentation à la suite de ce
      changement (fichier README, etc.)?

### CODE REVIEW

File: app/dao/services_dao.py [hunk 1]
```diff
@@ -186,12 +189,16 @@ def dao_fetch_live_services_data():
...
+        service_cache = redis_store.get(service_cache_key(service_id))
+        if service_cache:
+            service_cache_decoded = json.loads(service_cache.decode("utf-8"))["data"]
+            return Service.from_json(service_cache_decoded), service_cache_decoded
 
+    query = Service.query.filter_by(id=service_id).options(joinedload("users"))
     if only_active:
         query = query.filter(Service.active)
-
     return query.one()
```
Comment (sastels) [inline app/dao/services_dao.py:202]: Should we save to redis here if we had to query the db?

File: app/dao/templates_dao.py [hunk 2]
```diff
@@ -91,8 +94,26 @@ def dao_get_template_by_id_and_service_id(template_id, service_id, version=None)
...
+        if template_cache:
+            template_cache_decoded = json.loads(template_cache.decode("utf-8"))["data"]
+            if version:
+                return TemplateHistory.from_json(template_cache_decoded), template_cache_decoded
+            else:
+                return Template.from_json(template_cache_decoded), template_cache_decoded
+
+    elif version is not None:
         return TemplateHistory.query.filter_by(id=template_id, version=version).one()
     return Template.query.filter_by(id=template_id).one()
```
Comment (sastels) [inline app/dao/templates_dao.py:118]: same question re saving to redis here

**Review by sastels — APPROVED** [review]
Lots of tests 👍 

Ran everything locally, looks good. 

Just the questions around saving back to redis if there's a miss and we have to query the db.

File: app/dao/templates_dao.py [hunk 3]
```diff
@@ -91,8 +94,26 @@ def dao_get_template_by_id_and_service_id(template_id, service_id, version=None)
...
+    template_id, version=None, use_cache=False
+) -> Union[Union[Template, TemplateHistory], Tuple[Union[Template, TemplateHistory], dict]]:
+
+    if use_cache:
+        # When loading a SQLAlchemy object from cache it is in the transient state.
+        # We do not add it to the session. This would defeat the purpose of using the cache.
+        # The reasoning being that if the object is added to the session it would be queried
+        # to retrieve the latest data for that row. Since we do not want to add the object
+        # to the session it means some fields such as reply_to_text would be missing
+        # so we also return the cached data.
```
Comment (jimleroyer) [inline app/dao/templates_dao.py:107]: Great comment!

File: app/models.py [hunk 4]
```diff
@@ -567,7 +567,18 @@ def from_json(cls, data):
...
+        fields["go_live_user_id"] = fields.pop("go_live_user", None)
+        fields.pop("safelist", None)
+        fields.pop("permissions", None)
+        fields.pop("service_callback_api", None)
+        fields.pop("users", None)
+        fields.pop("annual_billing", None)
+        fields.pop("inbound_api", None)
+        fields.pop("letter_logo_filename", None)
+        fields.pop("letter_contact_block", None)
+        fields.pop("email_branding", None)
```
Comment (jimleroyer) [inline app/models.py:581]: Are we modifying current behavior by removing these properties from the service? Are these usually populated with the lazy loaders and not actual properties? (hence kinda making these useless on JSON translation)

File: app/notifications/process_notifications.py [hunk 5]
```diff
@@ -78,7 +78,6 @@ def persist_notification(
         template_version=template_version,
         to=recipient,
         service_id=service.id,
-        service=service,
```
Comment (jimleroyer) [inline app/notifications/process_notifications.py:81]: That was useless I guess? as we have the service ID already?

File: app/schemas.py [hunk 6]
```diff
@@ -65,6 +65,7 @@ def _validate_datetime_not_in_past(dte, msg="Date cannot be in the past"):
 
 class BaseSchema(ma.SQLAlchemyAutoSchema):  # type: ignore
     class Meta:
+        sqla_session = db.session
```
Comment (jimleroyer) [inline app/schemas.py:68]: Hmm why is it necessary?

File: tests/app/celery/test_tasks.py [hunk 7]
```diff
@@ -408,8 +410,9 @@ def test_should_process_all_sms_job(sample_job_with_placeholdered_template, mock
     ],
 )
 def test_process_row_sends_save_task(
-    notify_api, template_type, research_mode, expected_function, expected_queue, api_key_id, sender_id, mocker
+    notify_api, template_type, research_mode, expected_function, expected_queue, api_key_id, sender_id, mocker, notify_db_session
 ):
+    service_allowed_to_send_to_mock = mocker.patch("app.service.utils.safelisted_members", return_value=None)
```
Comment (jimleroyer) [inline tests/app/celery/test_tasks.py:415]: Hmm interesting, I didn't expect this in the PR.

File: tests/app/celery/test_tasks.py [hunk 8]
```diff
@@ -455,6 +459,87 @@ def test_process_row_sends_save_task(
     )
 
 
+def test_should_not_save_sms_if_restricted_service_and_invalid_number(notify_db_session, mocker):
```
Comment (jimleroyer) [inline tests/app/celery/test_tasks.py:462]: 👍

File: tests/app/celery/test_tasks.py [hunk 9]
```diff
@@ -455,6 +459,87 @@ def test_process_row_sends_save_task(
...
+        ),
+        template,
+        job,
+        service,
+    )
+
+    assert not save_sms_mock.called
+
+
+def test_should_not_save_sms_if_team_key_and_recipient_not_in_team(notify_db_session, mocker):
```
Comment (jimleroyer) [inline tests/app/celery/test_tasks.py:488]: 👍

File: tests/app/celery/test_tasks.py [hunk 10]
```diff
@@ -455,6 +459,87 @@ def test_process_row_sends_save_task(
...
+        ),
+        template,
+        job,
+        service,
+    )
+
+    assert not save_sms_mock.called
+
+
+def test_should_not_save_email_if_restricted_service_and_invalid_email_address(notify_db_session, mocker):
```
Comment (jimleroyer) [inline tests/app/celery/test_tasks.py:517]: ⚡

File: tests/app/invite/test_invite_rest.py [hunk 11]
```diff
@@ -190,7 +185,7 @@ def test_update_invited_user_for_wrong_service_returns_404(client, sample_invite
     assert json_response == "No result found"
 
 
-def test_update_invited_user_for_invalid_data_returns_400(client, sample_invited_user):
+def test_update_invited_user_for_invalid_data_returns_400(client, sample_invited_user, notify_db_session):
```
Comment (jimleroyer) [inline tests/app/invite/test_invite_rest.py:188]: Why is the `notify_db_session` necessary now? I think you alluded to it in a conversation last week..

**Review by jimleroyer — COMMENTED** [review]
Great test coverage! and great work..

My main concern like Steve is around not loading the service in the cache. But we can have a talk altogether tomorrow morning. Else I have a few questions here and there just to understand better the changes.

File: app/models.py (hunk 4 above)
Comment (jimleroyer) [reply app/models.py:581]: @Moro-Code confirmed that these are SQLAlchemy lazy loaders usually sitting in these properties, until properties are accessed.

**Pull request merged**